src/zone.c

```diff
@@ -115,89 +115,28 @@ diagnostic_push()
 msvc_diagnostic_ignored(4996)
 
 #if _WIN32
-nonnull_all
-static bool is_separator(int c)
-{
-  return c == '\\' || c == '/';
-}
-
-nonnull_all
-static bool is_rooted(const char *s)
-{
-  if ((s[0] >= 'A' && s[0] <= 'Z') || (s[0] >= 'a' && s[0] <= 'z'))
-    return s[1] == ':';
-  return false;
-}
-
-nonnull_all
-static bool is_relative(const char *s)
-{
-  // rooted paths can be relative, e.g. C:foo
-  if (is_rooted(s))
-    return !is_separator((unsigned char)s[2]);
-  if (is_separator((unsigned char)s[0]))
-    return !(s[1] == '?' || is_separator((unsigned char)s[1]));
-  return false;
-}
-
 // The Win32 API offers PathIsRelative, but it requires linking with shlwapi.
 // Rewriting a relative path is not too complex, unlike correct conversion of
 // Windows paths in general (https://googleprojectzero.blogspot.com/2016/02/).
 // Rooted paths, relative or not, unc and extended paths are never resolved
 // relative to the includer.
-nonnull((2,3))
-static int32_t resolve_path(
-  const char *includer, const char *include, char **path)
+nonnull_all
+static int32_t resolve_path(const char *include, char **path)
 {
-  // support relative non-rooted paths only
-  if (*includer && is_relative(include) && !is_rooted(include)) {
-    assert(!is_relative(includer));
-    char buffer[16];
-    int length = snprintf(
-      buffer, sizeof(buffer), "%s/%s", includer, include);
-    if (length < 0)
-      return ZONE_OUT_OF_MEMORY;
-    char *absolute;
-    if (!(absolute = malloc(length + 1)))
-      return ZONE_OUT_OF_MEMORY;
-    (void)snprintf(
-      absolute, (size_t)length + 1, "%s/%s", includer, include);
-    *path = _fullpath(NULL, absolute, 0);
-    free(absolute);
-  } else {
-    *path = _fullpath(NULL, include, 0);
-  }
-
-  if (*path)
+  if ((*path = _fullpath(NULL, include, 0)))
     return 0;
   return (errno == ENOMEM) ? ZONE_OUT_OF_MEMORY : ZONE_NOT_A_FILE;
 }
 #else
 nonnull_all
-static int32_t resolve_path(
-  const char *includer, const char *include, char **path)
+static int32_t resolve_path(const char *include, char **path)
 {
   char *resolved;
   char buffer[PATH_MAX + 1];
 
-  if (*includer && *include != '/') {
-    assert(*includer == '/');
-    int length = snprintf(
-      buffer, sizeof(buffer), "%s/%s", includer, include);
-    if (length < 0)
-      return ZONE_OUT_OF_MEMORY;
-    char *absolute;
-    if (!(absolute = malloc((size_t)length + 1)))
-      return ZONE_OUT_OF_MEMORY;
-    (void)snprintf(
-      absolute, (size_t)length + 1, "%s/%s", includer, include);
-    resolved = realpath(absolute, buffer);
-    free(absolute);
-  } else {
-    resolved = realpath(include, buffer);
-  }
+  resolved = realpath(include, buffer);
```

Comment:
I might be missing something, but you seem to be calling realpath twice? Could it be this is a left-over before the assignment was folded into the if-clause below?

Reply:
Seems I was a little to quick for my own good. Definitely an oversight. Thanks :+1: